New Admin Page: My Jetpack (connection management, rebased)

dereksmart opened the pull request:
Replay of #2252

Note: this PR does not include the dotcom changes that will be made on master user switch. That will be done in a following PR.

On this page, users will be able to view their connection information, as well as the ability to transfer "site ownership" aka master user.

You will need to run grunt locally to see the css file changes.

Prior to testing, please disconnect jetpack entirely.

Begin testing:
Set up 3 separate user accounts; (at least) Two admins and one editor (or any non-admin). I found it's easiest to log in to all three accounts in different windows/browsers.

As the first admin:

Connect Jetpack to your wpcom account. This will assign you as the master user.
As the second admin:

Log in, click on "My Jetpack": You should be able to see the primary connection holder, and a CTA to link the site to your wordpress.com account
Link your account to a unique wpcom account from the master.
Look at "My Jetpack" again. You should now be able to switch the primary account to your connection. You can also unlink your account.
As the non-admin: 
Repeat steps from the second admin. The Only difference should be that you can't do/see anything but link/unlink your own wpcom account to the site.

Some notes about My Jetpack

Only admins of the site will be able to disconnect the Jetpack site entirely.
You may now switch Master user roles
Only able to assign users as master who are both Admins and linked to a dotcom account.
You should only see yourself and the master user's info at any given time. If you are the master user, you will only see yourself.
Users below admin level still have access to the page, however they can only see their own connection status.
You should only be able to see the page if the site is connected.

dereksmart committed:
My Connection Page: Add initial functionality

adds the Jetapck connection information about the logged in user and the master user to My Connection page
- Admins can disconnect the Jetpack site, or their own link to wpcom
- Will display who is the Master user for a site
- Admins who are not master user will be able to set themselves as master
- Other users can link/unlink their accounts

New file: _inc/jp-connection.js
New file: class.jetpack-my-jetpack-page.php
New file: my-jetpack-page.php

jeffgolenski committed:
Style modification + converting html comments into php comments
Restructuring HTML + styling connection details
adding html + styles to connection page
responsive styling for connection page
More detailed styling for connection actions
formatting scss for best performance
Header nav btn adjustments for small mobile devices.
adding ellipsis overflow to usernames that are longer than a specified
length.
adjusting max width of usernames

jeffgolenski committed:
New style modifications
styling changes to connection 'tables'
Finding the best way to layout page

jeffgolenski committed:
adjusting layout for mobile

jeffgolenski committed:
style mods to last round of logic changes by @dereksmart
Cleaning up unneeded styles
more style refinements
Style adjustments for small mobile

jeffgolenski committed:
styling connection for single admin

jessefriedman committed:
Removed the disconnect and unlink links from footer

We want to funnel user traffic to disconnect their site to the “My
Connection” page so we can provide them with additional options and
support before they disconnect

jessefriedman committed:
Changed My Connection to be labeled My Jetpack

jessefriedman committed:
Improved disconnect flow and removed alert window

When a user wants to disconnect their site they will now see a message
that gives them the opportunity to first seek support.

dereksmart committed:
My Connection is now My Jetpack

file/element refactor to now call the page "My Jetpack" instead of My Connection

dereksmart committed:
My Jetpack: Log entry when switching master user

dereksmart committed:
My Jetpack: Refactor naming conventions & remove DRYness

Refactor to remove some DRYness, better naming conventions, better organization.

Created an additional function to handle the current user logic and separated it from the master user logic. It now is easier to read.

dereksmart committed:
My Jetpack: Clean up unused code

dereksmart committed:
Rebase css files

dereksmart committed:
Resolve merge conflicts

dereksmart committed:
Fix jshint errors

dereksmart committed:
Remove unused file

dereksmart committed:
Translate all the things

dereksmart committed:
My Jetpack: fix logic to check for other linked admins

dereksmart committed:
My Jetpack: Check to see if master user has been deleted

New function to determine if master user has been deleted/or is no longer a user on the site.

dereksmart committed:
My Jetpack: Add a success message on switching master user

jessefriedman committed:
Stats tracking on My Jetpack

We are now tracking
- My Jetpack views
- requests to change primary user
- successful primary user changes
- request to disconnect site
- confirmed disconnects
- users who request support instead of confirming disconnect

dereksmart committed:
My Jetpack: Don't show nav item if in dev mode

Apply "current" class on tab if active

dereksmart committed:
My Jetpack: UI consistency and improvements

Consistent column naming and remove unecessary column headers

dereksmart committed:
My Jetpack: Refactor UI if no other admins linked
- No more button
- No need for a custom table if no other admins

jeffgolenski committed:
adjusted breakpoint to fit header

adjusted breakpoint to fit header better on small mobile devices.

jeffgolenski committed:
fixed color of footer clouds

fixed color of footer clouds to match better on very tall screens

jeherve commented:
o Is there a way to stick the footer at the bottom of the page, to avoid layout issues like this one?

![screen shot 2015-06-18 at 11 07 10 am](https://cloud.githubusercontent.com/assets/426388/8227794/3f5dcb32-15aa-11e5-8ee0-68a1b329556c.png)

------

o I think there might be an issue with how we cache the data that is displayed.

Steps to reproduce:

1. Connect master user in browser 1
2. In another browser, connect another administrator
3. In a third browser, connect another user (an editor, for example)
3. Refresh the "My Jetpack" page in browser 1 and 2:
    - In browser 1, only the master user appears
    - In browser 2, only the master user and our own secondary admin appear

In browser 1 (Master user), I can add as many other users as I want, they don't appear.

-------

o Because of the Gravatars, the alignment of the usernames is a bit off

![screen shot 2015-06-18 at 11 11 52 am](https://cloud.githubusercontent.com/assets/426388/8227878/e54cf27a-15aa-11e5-8510-400ebc41cee4.png)

-----

o Can we add a bit of padding on the right of the Gravatar, so the username isn't too close to the image?

------

o Another small alignment issue: "Confirm Disconnect" and "Disconnect Site from WordPress.com" are not aligned vertically:

![screen shot 2015-06-18 at 11 23 20 am](https://cloud.githubusercontent.com/assets/426388/8228109/8b0071d2-15ac-11e5-8a63-8b1bdb60af2d.png)


------

o What's the difference between  "Confirm Disconnect" and "Disconnect Site from WordPress.com"? As a user I would be confused I think.

--------

o I wonder if links would make things easier to comprehend for people who're a bit confused with the differences between w.org and wp.com. We could add a link on both usernames, pointing to `wp-admin/profile.php` and `https://wordpress.com/me/account`

-----

o I don't know if it's immediately clear that the first row of the "My Jetpack" page is my own account, while the other rows are for other users. I wonder if we should change the wording of the first row to add "My" before "Site Username" and "WordPress.com username"?

![screen shot 2015-06-18 at 11 17 48 am](https://cloud.githubusercontent.com/assets/426388/8228013/d40eb29a-15ab-11e5-80a7-75ae64a91002.png)

-----

o In the 2 first columns we use the username. In the last one, we use the public display name. I would stick with usernames everywhere to avoid issues like these:

![screen shot 2015-06-18 at 11 31 26 am](https://cloud.githubusercontent.com/assets/426388/8228255/9d3f0614-15ad-11e5-8754-7493708c5255.png)

-----

o I think I would change "I need Support" into "I need help"

jessefriedman committed:
moved "change primary" cta

The second row for primary user was causing confusion so now the
ability to change the primary user has been moved to an option outside
the user list.

jessefriedman committed:
improved styling and layout changes

renamed My Jetpack js file and a few more styling improvements

zinigor commented:
I'm going to leave comments as soon as I see something so I don't forget. So please excuse me for the stream of consciousness thing :)
First - the My Jetpack page has a strange "Disconnecting Jetpack" section:
![my-jetpack](https://cloud.githubusercontent.com/assets/374293/8250272/ae5ad8e2-167a-11e5-83f0-f3e7c59c49c9.png)
Is this supposed to be a modal window that shows up when you click the "Disconnect site from WordPress.com"?

zinigor commented:
Also, as you can see in the previous screenshot, the primary user dropdown only shows the name of the user, not the username. Perhaps we should include it so that it's more clear who's who? I'm not sure it can be done, but the primary user we can highlight using bold letters instead of `(primary)`

zinigor commented:
Unlinking your account from WordPress.com does not ask for confirmation nor offers an undo. I'd say we need a simple "Are you sure?" dialog.

zinigor commented on _inc/jp-my-jetpack.js:
```diff
@@ -0,0 +1,69 @@
+     */
+    function confirmJetpackDisconnect() {
+        $( '#jetpack-disconnect' ).click( function() {
+            $( '#jetpack-disconnect-content' ).show();
```
We probably need the same thing somewhere on initialization, but with `hide()`, otherwise the confirmation block appears right away.

dereksmart replied:
this is moot when cs/js is compiled

zinigor commented on _inc/jp-my-jetpack.js:
```diff
@@ -0,0 +1,69 @@
+		$( '#jetpack-disconnect-content #confirm-disconnect' ).click( function() {
+			//Log My Jetpack event "confirm the disconnecting of a the site" in MC Stats
+			new Image().src = data.stats_urls.confirm_disconnect;
+		});
```
What do you think about adding a "Cancel" element here? It doesn't have to be a button, perhaps just a small link in case a user has changed their mind and wants to get back to the "My Jetpack" page.

dereksmart replied:
I like it. Maybe just a small `X` or `nevermind` link

dereksmart replied:
@jessefriedman fixed this in 1c5c2b04c37d0d13faaedcb7f8f4b266f02face1

zinigor commented:
In order to make the UI more consistend, what do you think about having the disconnection confirmation screen appear in a modal window just as the one we use for module information?

dereksmart commented:
> First - the My Jetpack page has a strange "Disconnecting Jetpack" section:
It looks like you haven't run `grunt` yet.  

> Use username instead of first name
Good idea :) 

>  but the primary user we can highlight using bold letters instead of (primary)
I personally like the word (primary) instead of bold for two reasons: 
- Sort of ambiguous 
- Accessibility/screen readers

> Asking for confirmation to unlink
Not a bad idea, What do you think @jessefriedman ?

zinigor commented:
Oh, right, `grunt`. :disappointed: :palm_tree: 
Disregard my comments about the disconnect block visible on first show. But I still think that a modal would serve better.

zinigor commented on _inc/lib/admin-pages/class.jetpack-my-jetpack-page.php:
```diff
@@ -0,0 +1,181 @@
+// Builds the My Jetpack page
+class Jetpack_My_Jetpack_Page extends Jetpack_Admin_Page {
+	// Show the settings page only when Jetpack is connected or in dev mode
+	protected $dont_show_if_not_active = true;
```
Nitpick, but maybe we should rename this to `hide_if_inactive`? Two negations in one variable name is kinda too much :)

dereksmart replied:
I agree, but should be changed in another PR because it's legacy Jetpack code

jessefriedman replied:
@zinigor I disagree that the disconnect confirmation should be in a modal. I think we need to start looking at getting away from modals, they provide added complication for mobile and accessibility. :)

dereksmart committed:
My Jetpack: Display username in dropdown instead of Full Name

jessefriedman committed:
Users can now exit disconnect screen

with the new disconnect view users are asked to confirm, if they need
support and now optionally can just close the screen. Also added
tracking for cancel button

zinigor commented on _inc/lib/admin-pages/class.jetpack-my-jetpack-page.php:
```diff
@@ -0,0 +1,181 @@
+	 * @return bool - True = Master user is still here, False = They're missing
+	 */
+	function jetpack_is_master_user_here() {
+		$master = Jetpack_Options::get_option( 'master_user' );
```
I can see that this method is being used in several places, but I don't like duplicating this option query here. Perhaps we can make this method return actual user data if it's there instead of just returning a boolean?

dereksmart replied:
I could just throw this logic into `jetpack_master_user_data()` and use that method to check if they are here?

zinigor replied:
Yes, I guess that would be perfect for now, and then we'll implement errors/warnings/exceptions/whatever for the case when there is no master user.

zinigor commented:
@dereksmart other than that I have said in the comments, everything else works fine. Tested using your test case with some additional tests like narrow screens, different browsers. Looks good!

dereksmart committed:
My Jetpack: Consolidate DB calls

ATM, we don't need a separate function to determine if the master user is missing.  Moved that logic into an existing method, and am using that to determine if master is here.

it succeeded in eliminating a DB call

dereksmart committed:
jp-my-jetpack.js: indent with tabs

dereksmart commented:
@jeherve addressing your points 
- Is there a way to stick the footer at the bottom of the page at all times?
Yep! it will require a minor refactor to the way we display all admin pages, so should be done in another PR

- We could add a link on both usernames
Sure, I would like to do this in a following PR as well.  FWIW the gravatar already links to your wporg profile.  

@jessefriedman will speak on verbiage of "I need Support"

jessefriedman commented:
@jeherve I prefer the word "Support" over the word "Help".  The goal here is to get them to support so they can work with your team to solve their problems. Clicking "I need help" could mean I'm being brought to FAQs, a support page with documentation, etc... It's a bit ambiguous and leaves it open to interpretation, so my vote is to keep it support so the user knows they are on a path to getting direct person to person support.

jessefriedman commented:
@zinigor regarding your question about a confirmation before unlinking an account (not disconnecting the site). I think it's something we should visit later. We don’t currently ask for confirmation to unlink. If we do add a confirmation modal it’ll be against the design pattern we just created for the disconnect site confirmation message. If we try to replicate what we did for the site disconnect it will take a while and I'd rather this this merged for 3.6 rather than hold it up any more :)

zinigor commented:
@jessefriedman OK, I think I agree here, furthermore, we have the big green button that allows you to reconnect, which kinda serves as undo. Let's try to polish this in the next take.

dereksmart commented:
All remaining comments are enhancements/preferences that can be addressed in future PRs.  Merging as-is.
